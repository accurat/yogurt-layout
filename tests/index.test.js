const { makeLayout } = require('../src/index')
const _ = require('lodash')

function removeNodeProperties(layout) {
  return _.mapValues(layout, ({ node, ...block }) => block)
}

it(`works with a simple layout`, () => {
  const layout = makeLayout({
    id: 'root',
    direction: 'column',
    width: 500,
    height: 500,
    padding: [10, 20, 30, 20],
    children: [
      { id: 'title', width: '100%', height: 50 },
      { id: 'content', width: '100%', height: 'auto' },
      { id: 'footer', width: '100%', height: 50 },
    ],
  })

  expect(removeNodeProperties(layout)).toMatchSnapshot()
})

it(`works with an advanced layout`, () => {
  const layout = makeLayout({
    id: 'root',
    direction: 'column',
    width: 500,
    height: 500,
    padding: 20,
    children: [
      { id: 'title', width: '100%', height: 50 },
      {
        id: 'chart',
        width: '100%',
        height: 'auto',
        direction: 'row',
        padding: 0,
        children: [
          { id: 'left', width: 100, height: '100%' },
          {
            id: 'center-wrapper',
            width: 'auto',
            height: '100%',
            padding: [10, 20],
            direction: 'row',
            children: [{ id: 'center', width: '100%', height: '100%' }],
          },
          { id: 'right', width: 100, height: '100%' },
        ],
      },
      { id: 'legend', width: '100%', height: 150 },
    ],
  })

  expect(removeNodeProperties(layout)).toMatchSnapshot()
})

it(`works with a double-auto layout`, () => {
  const layout = makeLayout({
    id: 'root',
    direction: 'row',
    width: 500,
    height: 500,
    children: [
      { id: 'aside', width: 100, height: '100%' },
      { id: 'content-1', width: 'auto', height: '100%' },
      { id: 'content-2', width: 'auto', height: '100%' },
    ],
  })

  expect(removeNodeProperties(layout)).toMatchSnapshot()
})

it(`errors out when layout has no space`, () => {
  expect(() =>
    makeLayout({
      id: 'root',
      direction: 'column',
      width: 500,
      height: 500,
      children: [
        { id: 'content-1', width: 'auto', height: 500 },
        { id: 'content-2', width: 'auto', height: 1 },
      ],
    })
  ).toThrow(`Block heights are overflowing! 500+1 > 500`)
})
