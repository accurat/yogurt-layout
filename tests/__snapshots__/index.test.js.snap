// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works with a double-auto layout 1`] = `
Object {
  "aside": Object {
    "bottom": 500,
    "height": 500,
    "id": "aside",
    "left": 0,
    "right": 100,
    "top": 0,
    "width": 100,
  },
  "content-1": Object {
    "bottom": 500,
    "height": 500,
    "id": "content-1",
    "left": 100,
    "right": 300,
    "top": 0,
    "width": 200,
  },
  "content-2": Object {
    "bottom": 500,
    "height": 500,
    "id": "content-2",
    "left": 300,
    "right": 500,
    "top": 0,
    "width": 200,
  },
  "root": Object {
    "children": Array [
      Object {
        "height": "100%",
        "id": "aside",
        "width": 100,
      },
      Object {
        "height": "100%",
        "id": "content-1",
        "width": "auto",
      },
      Object {
        "height": "100%",
        "id": "content-2",
        "width": "auto",
      },
    ],
    "direction": "row",
    "height": 500,
    "id": "root",
    "width": 500,
  },
}
`;

exports[`works with a simple layout 1`] = `
Object {
  "content": Object {
    "bottom": 420,
    "height": 360,
    "id": "content",
    "left": 20,
    "right": 480,
    "top": 60,
    "width": 460,
  },
  "footer": Object {
    "bottom": 470,
    "height": 50,
    "id": "footer",
    "left": 20,
    "right": 480,
    "top": 420,
    "width": 460,
  },
  "root": Object {
    "children": Array [
      Object {
        "height": 50,
        "id": "title",
        "width": "100%",
      },
      Object {
        "height": "auto",
        "id": "content",
        "width": "100%",
      },
      Object {
        "height": 50,
        "id": "footer",
        "width": "100%",
      },
    ],
    "direction": "column",
    "height": 500,
    "id": "root",
    "padding": Array [
      10,
      20,
      30,
      20,
    ],
    "width": 500,
  },
  "title": Object {
    "bottom": 60,
    "height": 50,
    "id": "title",
    "left": 20,
    "right": 480,
    "top": 10,
    "width": 460,
  },
}
`;

exports[`works with an advanced layout 1`] = `
Object {
  "center": Object {
    "bottom": 320,
    "height": 240,
    "id": "center",
    "left": 140,
    "right": 360,
    "top": 80,
    "width": 220,
  },
  "center-wrapper": Object {
    "bottom": 330,
    "height": 260,
    "id": "center-wrapper",
    "left": 120,
    "right": 380,
    "top": 70,
    "width": 260,
  },
  "chart": Object {
    "bottom": 330,
    "height": 260,
    "id": "chart",
    "left": 20,
    "right": 480,
    "top": 70,
    "width": 460,
  },
  "left": Object {
    "bottom": 330,
    "height": 260,
    "id": "left",
    "left": 20,
    "right": 120,
    "top": 70,
    "width": 100,
  },
  "legend": Object {
    "bottom": 480,
    "height": 150,
    "id": "legend",
    "left": 20,
    "right": 480,
    "top": 330,
    "width": 460,
  },
  "right": Object {
    "bottom": 330,
    "height": 260,
    "id": "right",
    "left": 380,
    "right": 480,
    "top": 70,
    "width": 100,
  },
  "root": Object {
    "children": Array [
      Object {
        "height": 50,
        "id": "title",
        "width": "100%",
      },
      Object {
        "children": Array [
          Object {
            "height": "100%",
            "id": "left",
            "width": 100,
          },
          Object {
            "children": Array [
              Object {
                "height": "100%",
                "id": "center",
                "width": "100%",
              },
            ],
            "direction": "row",
            "height": "100%",
            "id": "center-wrapper",
            "padding": Array [
              10,
              20,
            ],
            "width": "auto",
          },
          Object {
            "height": "100%",
            "id": "right",
            "width": 100,
          },
        ],
        "direction": "row",
        "height": "auto",
        "id": "chart",
        "padding": 0,
        "width": "100%",
      },
      Object {
        "height": 150,
        "id": "legend",
        "width": "100%",
      },
    ],
    "direction": "column",
    "height": 500,
    "id": "root",
    "padding": 20,
    "width": 500,
  },
  "title": Object {
    "bottom": 70,
    "height": 50,
    "id": "title",
    "left": 20,
    "right": 480,
    "top": 20,
    "width": 460,
  },
}
`;
